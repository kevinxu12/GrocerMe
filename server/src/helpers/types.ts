/**
 * @file Custom Types
 */
import User from '@src/models/User';
import { ItemRequestStatus, RequestStatus } from './model';
import { Types } from 'mongoose';
export interface CustomRequest extends Express.Request {
  app: any; // set to any for now
}

export interface LocationObject {
  description: string;
  place_id: string;
}
export interface ItemRequestParams {
  description: string;
  amount: number;
  title: string;
  location: LocationObject;
  date?: Date | null;
  image?: string | null;
  imageName?: string;
}
export interface ItemRequestParamsWithUser extends ItemRequestParams {
  user: User;
}

export interface ItemRequestFilters {
  status?: ItemRequestStatus;
}

export interface ItemMatchRequestFilters {
  status?: RequestStatus;
}

export interface MatchParams {
  itemRequestId: Types.ObjectId;
  amount: number;
  requester: User;
}

export interface SearchFilters {
  query: string;
  pageSize?: number;
  page?: number;
  additionalParams?: any;
}
