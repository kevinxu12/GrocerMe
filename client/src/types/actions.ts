export type ActionType = String;
export interface Action<P> {
  type: ActionType;
  payload: P;
}

export interface ChangeUsernamePayload {
  username: String;
}

export type ChangeUsernameAction = Action<ChangeUsernamePayload>;

export type LogoutAction = Action<{}>;

export type AuthAction = LogoutAction | ChangeUsernameAction;
