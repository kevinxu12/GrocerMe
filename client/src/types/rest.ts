/**
 * @file Backend types for the types in the REST API.
 * These will probably grow stale as the api's change, so maye we should think of a better way to keep the frontend and backend types synced
 * @author Kevin Xu
 */
export interface Role {
  _id?: string;
  status: boolean;
  code: RoleCode;
}

export interface User {
  _id: string;
  firstTime: boolean;
  roles: Role[];
  name: string;
  email: string;
}

export interface SupplierRequest {
  _id: string;
  requester?: User;
  status: string;
  createdAt: Date;
  updatedAt: Date;
  active: boolean;
  email: string;
  approvedBy?: string;
}

export interface SuccessResponse<T> {
  statusCode: string;
  message: string;
  data: T;
}

export enum StatusCode {
  SUCCESS = '10000',
  FAILURE = '10001',
  RETRY = '10002',
  INVALID_ACCESS_TOKEN = '10003',
}

export enum RoleCode {
  CONSUMER = 'CONSUMER',
  SUPPLIER = 'SUPPLIER',
  ADMIN = 'ADMIN',
}

export interface LocationObject {
  description: string;
  place_id: string;
}

export enum RequestStatus {
  REJECTED = 'REJECTED',
  ACCEPTED = 'ACCEPTED',
  AWAITING = 'AWAITING',
}

export enum ItemRequestStatus {
  REJECTED = 'REJECTED',
  ACCEPTED = 'ACCEPTED',
  AWAITING = 'AWAITING',
  SOLD = 'SOLD',
}
export interface ItemRequest {
  _id: string;
  requester: User; // we make this optional because it makes testing easier.
  status: ItemRequestStatus;
  createdAt: Date;
  updatedAt: Date;
  active: boolean;
  email: string;
  amount: number;
  description: string;
  location: LocationObject;
  title: string;
  approvedBy?: string;
  amountSold: number;
  imageUrl?: string;
}

export interface ItemMatchRequest {
  _id: string;
  requester: User; // we make this optional because it makes testing easier.
  status: RequestStatus;
  createdAt: Date;
  updatedAt: Date;
  requesterEmail: string;
  supplierEmail: string;
  item: ItemRequest;
  amount: number;
}
