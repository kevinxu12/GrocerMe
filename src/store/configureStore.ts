import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';

import { createReducer } from './reducers';
import thunk from 'redux-thunk';

export function configureAppStore(initialState = {}) {
  /* eslint-disable */
  let composeEnhancers = window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose || compose;
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [sagaMiddleware, thunk];

  const enhancers = [applyMiddleware(...middlewares)];
  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers),
  );

  // const reduxSagaMonitorOptions = {};
  // const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
  // const { run: runSaga } = sagaMiddleware;

  // // Create the store with saga middleware
  // const middlewares = [sagaMiddleware];

  // const enhancers = [
  //   createInjectorsEnhancer({
  //     createReducer,
  //     runSaga,
  //   }),
  // ] as StoreEnhancer[];

  // const store = configureStore({
  //   reducer: createReducer(),
  //   middleware: [...getDefaultMiddleware(), ...middlewares],
  //   devTools:
  //     /* istanbul ignore next line */
  //     process.env.NODE_ENV !== 'production' ||
  //     process.env.PUBLIC_URL.length > 0,
  //   enhancers,
  // });

  return store;
}
