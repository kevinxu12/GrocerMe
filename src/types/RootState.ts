// [IMPORT NEW CONTAINERSTATE ABOVE] < Needed for generating containers seamlessly

/* 
  Because the redux-injectors injects your reducers asynchronously somewhere in your code
  You have to declare them here manually
*/
interface HomeState {
  username: String;
}
export interface RootState {
  home: HomeState; // [INSERT NEW REDUCER KEY ABOVE] < Needed for generating containers seamlessly
}
