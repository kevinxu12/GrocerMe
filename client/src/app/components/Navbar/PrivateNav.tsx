/**
 * @file Component for Navbar shown for logged-in users
 * @author Kevin Xu
 */
import React, { ComponentType, memo, useState } from 'react';
import { ProSidebar, Menu, MenuItem } from 'react-pro-sidebar';
import * as FaIcons from 'react-icons/fa';
import * as AiIcons from 'react-icons/ai';
import styled from 'styled-components/macro';
import './custom.scss';
import { connect } from 'react-redux';
import { compose } from '@reduxjs/toolkit';
import { logoutWithThunk } from 'store/auth/thunk';
import { GenericThunkDispatch } from 'types/actions';

interface PrivateNavDispatchType {
  onLogout: () => void;
}
interface PrivateNavPropsType extends PrivateNavDispatchType {}

/**
 * @param {object} props props passed
 * @param {Function} props.onLogout logout function
 * @returns {React.ElementType} Private Navbar Component
 */
const PrivateNav = (props: PrivateNavPropsType): React.ReactElement => {
  const [sidebar, setSidebar] = useState(false);
  /**
   * @returns {boolean} whether to show side bar or not
   */
  const showSidebar = () => setSidebar(!sidebar);
  const SidebarProps = {
    open: sidebar,
  };
  return (
    <Wrapper>
      <Icon>
        <FaIcons.FaBars onClick={showSidebar} />
      </Icon>
      <Sidebar {...SidebarProps}>
        <ProSidebar>
          <Menu>
            <MenuItem icon={<AiIcons.AiOutlineClose />} onClick={showSidebar} />
            <MenuItem icon={<AiIcons.AiFillHome />}>Home</MenuItem>
            <MenuItem
              icon={<AiIcons.AiOutlineLogout />}
              onClick={props.onLogout}
            >
              Logout
            </MenuItem>
          </Menu>
        </ProSidebar>
      </Sidebar>
    </Wrapper>
  );
};

/**
 * Maps dispatch functions to component props
 *
 * @param {GenericThunkDispatch} dispatch the dispatch object from redux
 * @returns {PrivateNavDispatchType} Object passed to props containing redux dispatch functions
 */
function mapDispatchToProps(
  dispatch: GenericThunkDispatch,
): PrivateNavDispatchType {
  return {
    /**
     * Dispatches a logout action
     */
    onLogout: () => {
      dispatch(logoutWithThunk());
    },
  };
}

const withConnect = connect(null, mapDispatchToProps);

export default compose<ComponentType>(withConnect, memo)(PrivateNav);

const Icon = styled.div`
  min-height: 5%;
  min-width: 5%;
`;

const Sidebar = styled.nav`
  display: none;
  ${({ open }: any) =>
    open &&
    `
    width: 20%;
    height: 100vh;
    display: flex;
    justify-content: center;
    position: fixed;
    top: 0;
    left: 0%;
    `}
`;

const Wrapper = styled.nav`
  display: flex;
  margin-right: -1rem;
  width: 100%;
`;
