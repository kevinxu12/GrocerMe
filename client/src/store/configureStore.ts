/**
 * @file Class configuring the redux store
 * @author Kevin Xu
 */
import { createStore, applyMiddleware, compose, CombinedState } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { createReducer } from './reducers';
import thunk from 'redux-thunk';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { RootState } from 'types/RootState';

/**
 * @param {object} initialState the initial state of the store. Default to an empty object
 * @returns {object} the redux store for this specific project
 */
export function configureAppStore(initialState = {}) {
  /* eslint-disable */
  let composeEnhancers =
    (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) ||
    compose;
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  // const reducer = storage.reducer(createReducer());
  const rootReducer = createReducer();
  // const engine = createEngine('my-save-key', reviver);
  // const middleware = storage.createMiddleware(engine);
  const middlewares = [sagaMiddleware, thunk]; //middleware];

  const persistConfig = {
    key: 'root',
    storage: storage,
  };

  const pReducer = persistReducer<CombinedState<RootState>, any>(
    persistConfig,
    rootReducer,
  );

  const enhancers = [applyMiddleware(...middlewares)];
  const store = createStore(
    pReducer,
    initialState,
    composeEnhancers(...enhancers),
  );
  
  return store;
}
