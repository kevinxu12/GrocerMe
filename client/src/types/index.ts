/**
 * @file Top-level entry type for the types directory
 * @author Kevin Xu
 */
import { RootState } from './RootState';
export type setMessageSnackbarType = (string) => void;
export type { RootState };
