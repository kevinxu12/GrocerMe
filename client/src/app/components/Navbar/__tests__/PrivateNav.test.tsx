/**
 * @file Testing Private Navbar component
 * @author Kevin Xu
 */
 import * as React from 'react';
 import { render } from '@testing-library/react';
 import { Provider } from 'react-redux';
 import configureStore from 'redux-mock-store';
 import PrivateNav from './../PrivateNav';
 import { Role, RoleCode } from 'types/rest';
 import { AuthState } from 'types/RootState';
 
 
 describe('<PrivateNav />', () => {
   const mockStore = configureStore();
   let store;
 
   it('should match snapshot for private nav for with supplier', () => {
     const TEST = 'test';
     const testRole: Role = { status: true, code: RoleCode.SUPPLIER};
     const loggedInState: AuthState = {username: TEST, roles: [testRole]};
     const state = { auth: loggedInState };
     store = mockStore(state);
     const loadingIndicator = render(
       <Provider store={store}>
         <PrivateNav />
       </Provider>,
     );
     expect(loadingIndicator.container.firstChild).toMatchSnapshot();
   });
 });