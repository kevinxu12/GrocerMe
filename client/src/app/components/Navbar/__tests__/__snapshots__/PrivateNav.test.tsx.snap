// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateNav /> should match snapshot for private nav for with supplier 1`] = `
.c1 {
  min-height: 5%;
  min-width: 5%;
}

.c2 {
  display: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: -1rem;
  width: 100%;
}

<nav
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      />
    </svg>
  </div>
  <nav
    class="c2"
  >
    <div
      class="pro-sidebar"
    >
      <div
        class="pro-sidebar-inner"
      >
        <div
          class="pro-sidebar-layout"
        >
          <nav
            class="pro-menu inner-submenu-arrows"
          >
            <ul>
              <li
                class="pro-menu-item"
              >
                <div
                  class="pro-inner-item"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="pro-icon-wrapper"
                  >
                    <span
                      class="pro-icon"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 1024 1024"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    class="pro-item-content"
                  />
                </div>
              </li>
              <li
                class="pro-menu-item"
              >
                <div
                  class="pro-inner-item"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="pro-icon-wrapper"
                  >
                    <span
                      class="pro-icon"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 1024 1024"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M946.5 505L534.6 93.4a31.93 31.93 0 0 0-45.2 0L77.5 505c-12 12-18.8 28.3-18.8 45.3 0 35.3 28.7 64 64 64h43.4V908c0 17.7 14.3 32 32 32H448V716h112v224h265.9c17.7 0 32-14.3 32-32V614.3h43.4c17 0 33.3-6.7 45.3-18.8 24.9-25 24.9-65.5-.1-90.5z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    class="pro-item-content"
                  >
                    Home
                  </span>
                </div>
              </li>
              <li
                class="pro-menu-item"
              >
                <div
                  class="pro-inner-item"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="pro-icon-wrapper"
                  >
                    <span
                      class="pro-icon"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 1024 1024"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M868 732h-70.3c-4.8 0-9.3 2.1-12.3 5.8-7 8.5-14.5 16.7-22.4 24.5a353.84 353.84 0 0 1-112.7 75.9A352.8 352.8 0 0 1 512.4 866c-47.9 0-94.3-9.4-137.9-27.8a353.84 353.84 0 0 1-112.7-75.9 353.28 353.28 0 0 1-76-112.5C167.3 606.2 158 559.9 158 512s9.4-94.2 27.8-137.8c17.8-42.1 43.4-80 76-112.5s70.5-58.1 112.7-75.9c43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 7.9 7.9 15.3 16.1 22.4 24.5 3 3.7 7.6 5.8 12.3 5.8H868c6.3 0 10.2-7 6.7-12.3C798 160.5 663.8 81.6 511.3 82 271.7 82.6 79.6 277.1 82 516.4 84.4 751.9 276.2 942 512.4 942c152.1 0 285.7-78.8 362.3-197.7 3.4-5.3-.4-12.3-6.7-12.3zm88.9-226.3L815 393.7c-5.3-4.2-13-.4-13 6.3v76H488c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h314v76c0 6.7 7.8 10.5 13 6.3l141.9-112a8 8 0 0 0 0-12.6z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    class="pro-item-content"
                  >
                    Logout
                  </span>
                </div>
              </li>
              <li
                class="pro-menu-item"
              >
                <div
                  class="pro-inner-item"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="pro-icon-wrapper"
                  >
                    <span
                      class="pro-icon"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 1024 1024"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M924.8 385.6a446.7 446.7 0 0 0-96-142.4 446.7 446.7 0 0 0-142.4-96C631.1 123.8 572.5 112 512 112s-119.1 11.8-174.4 35.2a446.7 446.7 0 0 0-142.4 96 446.7 446.7 0 0 0-96 142.4C75.8 440.9 64 499.5 64 560c0 132.7 58.3 257.7 159.9 343.1l1.7 1.4c5.8 4.8 13.1 7.5 20.6 7.5h531.7c7.5 0 14.8-2.7 20.6-7.5l1.7-1.4C901.7 817.7 960 692.7 960 560c0-60.5-11.9-119.1-35.2-174.4zM482 232c0-4.4 3.6-8 8-8h44c4.4 0 8 3.6 8 8v80c0 4.4-3.6 8-8 8h-44c-4.4 0-8-3.6-8-8v-80zM270 582c0 4.4-3.6 8-8 8h-80c-4.4 0-8-3.6-8-8v-44c0-4.4 3.6-8 8-8h80c4.4 0 8 3.6 8 8v44zm90.7-204.5l-31.1 31.1a8.03 8.03 0 0 1-11.3 0L261.7 352a8.03 8.03 0 0 1 0-11.3l31.1-31.1c3.1-3.1 8.2-3.1 11.3 0l56.6 56.6c3.1 3.1 3.1 8.2 0 11.3zm291.1 83.6l-84.5 84.5c5 18.7.2 39.4-14.5 54.1a55.95 55.95 0 0 1-79.2 0 55.95 55.95 0 0 1 0-79.2 55.87 55.87 0 0 1 54.1-14.5l84.5-84.5c3.1-3.1 8.2-3.1 11.3 0l28.3 28.3c3.1 3.1 3.1 8.1 0 11.3zm43-52.4l-31.1-31.1a8.03 8.03 0 0 1 0-11.3l56.6-56.6c3.1-3.1 8.2-3.1 11.3 0l31.1 31.1c3.1 3.1 3.1 8.2 0 11.3l-56.6 56.6a8.03 8.03 0 0 1-11.3 0zM846 582c0 4.4-3.6 8-8 8h-80c-4.4 0-8-3.6-8-8v-44c0-4.4 3.6-8 8-8h80c4.4 0 8 3.6 8 8v44z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    class="pro-item-content"
                  >
                     
                    Supplier Home
                     
                  </span>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
      <div
        aria-label="overlay"
        class="overlay"
        role="button"
        tabindex="0"
      />
    </div>
  </nav>
</nav>
`;
